src/mpid/ch4/generic/am/mpidig_am_recv.h

```diff
@@ -128,29 +129,81 @@ static inline int MPIDIG_do_irecv(void *buf, MPI_Aint count, MPI_Datatype dataty
+                MPIDI_Datatype_check_size(datatype, count, data_sz);
+                mpi_errno =
+                    MPIDIG_gpu_ipc_send(buf, count, datatype, data_sz, rank, tag, comm,
+                                        context_offset, rts_req, &unexp_req);
+                MPIR_ERR_CHECK(mpi_errno);
+
                 MPIDIG_REQUEST(unexp_req, req->rreq.match_req) = NULL;
-            } else {
-                /* Enqueuing path: CH4 already allocated a request.
-                 * Record the passed `*request` to `match_req` so that we can complete it
-                 * later when `unexp_req` completes.
-                 * See MPIDI_recv_target_cmpl_cb for actual completion handler. */
-                MPIDIG_REQUEST(unexp_req, req->rreq.match_req) = *request;
+                if (*request == NULL) {
+                    *request = unexp_req;
+                }
+                goto fn_exit;
+            } else
+#endif
+            {
+                /* Matching receive is now posted, tell the netmod/shmmod */
+                MPIR_Datatype_add_ref_if_not_builtin(datatype);
+                MPIDIG_REQUEST(unexp_req, datatype) = datatype;
+                MPIDIG_REQUEST(unexp_req, buffer) = (char *) buf;
+                MPIDIG_REQUEST(unexp_req, count) = count;
+                if (*request == NULL) {
+                    /* Regular (non-enqueuing) path: MPIDIG is responsbile for allocating
+                     * a request. Here we simply return `unexp_req` */
+                    *request = unexp_req;
+                    /* Mark `match_req` as NULL so that we know nothing else to complete when
+                     * `unexp_req` finally completes. (See below) */
+                    MPIDIG_REQUEST(unexp_req, req->rreq.match_req) = NULL;
+                } else {
+                    /* Enqueuing path: CH4 already allocated a request.
+                     * Record the passed `*request` to `match_req` so that we can complete it
+                     * later when `unexp_req` completes.
+                     * See MPIDI_recv_target_cmpl_cb for actual completion handler. */
+                    MPIDIG_REQUEST(unexp_req, req->rreq.match_req) = *request;
+                }
+                mpi_errno = MPIDIG_do_long_ack(unexp_req);
+                MPIR_ERR_CHECK(mpi_errno);
+                goto fn_exit;
             }
-            mpi_errno = MPIDIG_do_long_ack(unexp_req);
+        } else
+#ifdef MPL_HAVE_GPU
+        if (MPIDIG_REQUEST(unexp_req, req->status) & MPIDIG_REQ_GPU_IPC) {
+            void *src_buf;
+            MPL_gpu_ipc_mem_handle_t mem_handle = MPIDIG_REQUEST(unexp_req, req->ireq.mem_handle);
+            MPL_gpu_ipc_open_mem_handle(&src_buf, mem_handle);
+
+            MPI_Aint data_sz;
+            MPIDI_Datatype_check_size(datatype, count, data_sz);
+            yaksa_request_t yaksa_request;
+            mpi_errno =
+                yaksa_iunpack(src_buf, data_sz, buf, data_sz, YAKSA_TYPE__BYTE, 0, &yaksa_request);
```

Comment:
FIXME: should not do unpack here. Set request status and postpone upack to callback.